teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,428 @@
+
+3. Backwards Compatibility
+
+    This model [diverges](#http-interface) from the existing WebHookInterceptor model. For backwards compatibility, we can keep supporting the existing [WebhookInterceptor interface](https://github.com/tektoncd/triggers/blob/master/docs/eventlisteners.md#event-interceptor-services) for a few releases or until beta.
+
+4. Secrets and RBAC concerns 
+
+    Today, if an interceptor needs access to a secret (e.g GitHub payload validation), the service account for the EventListener needs to get access to secrets. With the proposed model, this is no longer the case. Instead the GitHub interceptor deployment should use a role that has read access to secrets. This role would have to be a cluster role if it needs access to secrets across multiple namespaces. A security conscious admin might want to use [resourceNames](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources) to lock down the list of secrets the deployment has access to.
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+```
+<<[UNRESOLVED @dibyom]>>
+<<[/UNRESOLVED]>>
+```
+
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+
+```go
+type InterceptorRequest struct {
+	// Body is the incoming HTTP event body
+	Body []byte `json:"body,omitempty"`
+	// Header are the headers for the incoming HTTP event
+	Header map[string][]string `json:"header,omitempty"`
+
+	// InterceptorParams are the user specified params for the interceptor
+	InterceptorParams map[string]string `json:"interceptor_ params,omitempty"`
+
+	Context *TriggerContext
+}
+
+type TriggerContext struct {
+	// EventUrl is the URL of the incoming event
+	EventUrl string `json:"url,omitempty"`
```

Comment:
I'm not sure yet @dibyom and @khrm if they are orthogonal or not

But some thought into how the new `TriggerContext` type here and the new Trigger CRD in https://github.com/tektoncd/community/blob/master/teps/0009-trigger-crd.md#trigger-crd reconcile might be warranted.

Even if in the end we assure everyone that they are orthogonal, some text around here explaining that would be helpful I think.

Conversely, if this `TriggerContext` needs to say be included in the new Trigger CRD spec (but at least for me though I need to think about that more to know if that would even be the case).

Perhaps a discussion topic for the next WG call?

Reply:
sure. I think they are pretty orthogonal. The TriggerID field in the context contains the namespace and name of the Trigger CRD. The other contents of the Context are either generated (the EventID) or from the incoming Event (EventUrl)

Reply:
Those field clarifications help me - thanks @dibyom 

Some additional brainstorming prior to the WG call ... we'll see what if anything sticks:
- would a timestamp in the TriggerContext help with debug ?
- would including a TriggerContext ref in the interceptor response potentially allow for some debugging help 
-- assist correlating of req / response
-- possibly facilitate access to the trigger CRD and its conditions to see if disruptions occurred during the processing of the interceptor request

Reply:
Per last WG discussion, all of these sound like useful additions but they do not have to be added in the initial implementation (i.e. can be part of a follow up)

Reply:
with that @dibyom could these be listed in the non-goals or alternatives sections (or maybe we should propose a new "follow up" section to the TEP template ;-) ) so we don't lose track of them?

Ultimately, I could MVP line items as github kind/features targeted to specific milestones , and follow up items as github kind/features left in the backlog (vs. added to a milestone) that could then go through the github lifecycle machinery already in place.

Thoughts?

Reply:
Good idea. I'll add a WorkPlan section; follow up issues/PRs will be linked to https://github.com/tektoncd/triggers/issues/271

Reply:
cool thanks

Reply:
Added a Work Plan section.